src/app/marketplace/edit-ad.component.ts

```diff
@@ -0,0 +1,329 @@
+  }
+
+  resolveMenu(data: AdBasicData) {
+    return this.authHelper.userMenu(data.owner, Menu.SEARCH_ADS);
+  }
+
+  /**
+   * Changes the current currency updating other referenced fields like price and promotional price
+   */
+  protected updateCurrency(id: string, data: AdBasicData) {
+    this.currency = data.currencies.find(c => c.id === id || c.internalName === id);
+  }
+
+  /**
+   * Creates a list of categories with extra information (e.g level, leaf) used for rendering
+   */
+  protected populateCategories(items: AdCategoryWithChildren[], level: number) {
+    for (const item of items) {
+      const hasChildren = !empty(item.children);
+      const hierarchy: HierarchyItem = item as any;
+      hierarchy.level = level;
+      hierarchy.leaf = !hasChildren;
+      this.categories.push(hierarchy);
+      if (hasChildren) {
+        this.populateCategories(item.children, level + 1);
+      }
+    }
+  }
+
+  /**
+   * Resolves the current ad status label
+   */
+  get status(): string {
+    return this.marketplaceHelper.resolveStatusLabel((this.data as AdDataForEdit).status);
+  }
+
+  /**
+   * Creates or updates the current ad and allows to create
+   * a new ad based on the current one
+   */
+  save(insertNew?: boolean) {
+
+    validateBeforeSubmit(this.form);
+    if (!this.form.valid) {
+      return;
+    }
+
+    const value = this.form.value;
+
+    value.publicationPeriod = this.ApiHelper.dateRange(value.publicationBeginDate, value.publicationEndDate);
+    delete value['publicationBeginDate'];
+    delete value['publicationEndDate'];
+
+    value.promotionalPeriod = this.ApiHelper.dateRange(value.promotionalBeginDate, value.promotionalEndDate);
+    delete value['promotionalBeginDate'];
+    delete value['promotionalEndDate'];
+
+    const onFinish: any = (id: string) => {
+      this.notification.snackBar(this.i18n.ad.adSaved);
+      if (insertNew) {
+        EditAdComponent.BASED_ON_ID = id || this.id;
+        if (this.create) {
+          this.reload();
+        } else {
+          this.router.navigate(['/marketplace', this.owner, this.data.kind, 'new'], { replaceUrl: true });
+        }
+      } else if (this.basedOnId) {
+        this.router.navigate(['/marketplace', 'edit', id], { replaceUrl: true });
+      } else {
+        history.back();
+      }
+    };
+
+    if (this.create) {
+
+      value.images = this.uploadedImages.map(i => i.id);
+
+      if (this.data.requiresAuthorization) {
+        // When requires authorization submit as draft when saving for first time
+        value.submitForAuthorization = false;
+      }
+
+      this.addSub(this.marketplaceService.createAd({
+        user: this.owner,
+        body: value
+      }).subscribe(onFinish));
+
+    } else {
+
+      const updateAdReq = this.marketplaceService.updateAd({
+        ad: this.id,
+        body: value
+      });
+
+      // If the main image has changed reload the ad version
+      if (this.mainImageChanged()) {
```

Comment:
I didn't completely understand what this does. Anyway, if we're going to change it to the 'full-ad' edit, I think this won't be needed anymore.

Reply:
When we handle images separatedly if the main ad image changes (the first one) the ad version is incremented, so we need to detect that case and retrieve the new ad version, otherwise it will fail with a version control exception when saving.